src/ripple/shamap/README.md

```diff
@@ -26,164 +37,307 @@ There are two different ways of building and using a SHAMap:
-constructed, later SHAMaps are usually created by calling
-snapShot(bool isMutable) on the original SHAMap().  The returned SHAMap
-has the expected characteristics (mutable or immutable) based on the passed
-in flag.
+constructed, later SHAMaps are usually created by calling snapShot(bool
+isMutable) on the original SHAMap().  The returned SHAMap has the expected
+characteristics (mutable or immutable) based on the passed in flag.
 
 It is cheaper to make an immutable snapshot of a SHAMap than to make a mutable
-snapshot.  If the SHAMap snapshot is mutable then any of the nodes that might
-be modified must be copied before they are placed in the mutable map.
+snapshot.  If the SHAMap snapshot is mutable then any of the nodes that might be
+modified must be copied before they are placed in the mutable map.
 
+A new SHAMap is created with each new ledger round.  Transactions not executed
+in the previous ledger populate the SHAMap for the new ledger.
 
-## SHAMap Thread Safety ##
+## Storing SHAMap data in the database ##
 
-*This description is obsolete and needs to be rewritten.*
+When consensus is reached, the ledger is closed.  As part of this process, the
+SHAMap is stored to the database by calling `SHAMap::flushDirty` with parameters
+which specify which nodes are to be written.
 
-SHAMaps can be thread safe, depending on how they are used.  The SHAMap
-uses a SyncUnorderedMap for its storage.  The SyncUnorderedMap has three
-thread-safe methods:
+`flushDirty` walks the SHAMap by calling `SHAMap::walkSubTree`.  This process
+must ensure that each node is only owned by this tree, and so performs a COW
+operation (`unshare()`) as it walks each node (from the root down).  This is
+done by the `preFlushNode` function.
 
- * size(),
- * canonicalize(), and
- * retrieve()
+For each inner node encountered (starting with the root node), each of the
+children are inspected (from 1 to 16).  For each child, if it has a non-zero
+sequence number (unshareable), the child is first COW'd.  Then if the child is
+an inner node, we recurse down to that node's children.  Otherwise we've found a
+leaf node and that node is written to the database.  A count of each leaf node
+that is visited is kept.  The hash of the data in the leaf node is computed at
+this time, and the child is reassigned back into the parent inner node just in
+case the COW operation created a new pointer to this leaf node.
 
-As long as the SHAMap uses only those three interfaces to its storage
-(the mTNByID variable [which stands for Tree Node by ID]) the SHAMap is
-thread safe.
+After processing each node, the node is then marked as sharable again by setting
+its sequence number to 0.
 
+After all of an inner node's children are processed, then it's hash is updated
+and the inner node is written to the database.  Then this inner node is assigned
+back into it's parent node, again in case the COW operation created a new
+pointer to it.
```

Comment:
When I read this paragraph, I have the following questions, can we consider adding more details:
1. At what point does the actual database write occur?
2. How does share/unshare work?
3. Can we give an overview of how flushDirty works? One aspect of it is how does the software keep track of what needs to be flushed, and by design, does it only walk the tree that is "dirty"?

Reply:
> 1. At what point does the actual database write occur?

For the inner node, which is the subject of this paragraph, I really don't know how to say it more clearly than the first sentence of this paragraph.  Suggestions are welcome.

> 2.  How does share/unshare work?

There is no explicit `share` function.  Nodes are marked as "sharable" by:

> After processing each node, the node is then marked as sharable again by setting
> its sequence number to 0.

When a snapshot of a SHAMap is made, nodes that are sharable are shared via the copy constructor of the `std::shared_ptr` which owns the node.

I've updated the paragraph that describes `flushDirty` to include `unshare()`, as both of these functions just call `walkSubTree`.  Following this paragraph is a detailed description of `walkSubTree`.

> 3. Can we give an overview of how flushDirty works? One aspect of it is how does the software keep track of what needs to be flushed, and by design, does it only walk the tree that is "dirty"?

`flushDirty` writes the entire tree to the database.  I've added a sentence to the `walkSubTree` description to say this.

Reply:
Thanks! I will resolve the comment